diffrax/_integrate.py

```diff
@@ -775,9 +775,51 @@ def _save_t1(subsaveat, save_state):
                 save_state = _save(tfinal, yfinal, args, subsaveat.fn, save_state)
         return save_state
 
-    save_state = jtu.tree_map(
-        _save_t1, saveat.subs, final_state.save_state, is_leaf=_is_subsaveat
+    def _save_ts_impl(ts, fn, _save_state):
+        def _cond_fun(__save_state):
+            return __save_state.saveat_ts_index < len(_save_state.ts)
+
+        def _body_fun(__save_state):
+            idx = __save_state.save_index
+            ts = __save_state.ts.at[idx].set(t0)
+            ys = jtu.tree_map(
+                lambda _y, _ys: _ys.at[idx].set(_y),
+                fn(t0, yfinal, args),
+                __save_state.ys,
+            )
+            return SaveState(
+                saveat_ts_index=idx + 1,
+                ts=ts,
+                ys=ys,
+                save_index=idx + 1,
+            )
+
+        return inner_while_loop(
+            _cond_fun,
+            _body_fun,
+            _save_state,
+            max_steps=len(ts),
+            buffers=_inner_buffers,
+            checkpoints=len(ts),
+        )
+
+    def _save_ts(subsaveat: SubSaveAt, save_state: SaveState) -> SaveState:
+        if subsaveat.ts is not None:
+            save_state = _save_ts_impl(subsaveat.ts, subsaveat.fn, save_state)
+        return save_state
+
+    # if t0 == t1 then we don't enter the integration loop. In this case we have to
+    # manually update the saved ts and ys if we want to save at "intermediate"
+    # times specified by saveat.subs.ts
+    save_state = jax.lax.cond(
+        t0 == t1,
+        lambda _save_state: jtu.tree_map(
+            _save_ts, saveat.subs, _save_state, is_leaf=_is_subsaveat
+        ),
```

Comment:
I think this `t0 == t1` branch can be made much more efficient: I imagine you should be able to do the tree-mapped equivalent of `ys = jnp.where(ts == t0, y0, jnp.inf)`. Does something go wrong with this approach that I'm missing?

(Regardless of the above, do also note that you are doing a tree-map-of-a-loop. If you can, it is usually much more efficient to do a loop-of-a-tree-map.)

Reply:
Thanks for the note about tree-map-of-a-loop vs. loop-of-a-tree-map! This was a gotcha I was unaware of.

I am probably missing something here but it seems to me at least naively that things could get screwed up based on the way data is saved in SaveState. In my test, we would have for `saveat`
```python
SaveAt(
  subs=(
    SubSaveAt(t0=False, t1=True, ts=f32[3], steps=False, fn=<function save_y>),
    SubSaveAt(t0=True, t1=False, ts=f32[3], steps=False, fn=<function save_y>)
  ),
  dense=False,
  solver_state=False,
  controller_state=False,
  made_jump=False
)
``` 
and for `final_state.save_state` we have 
```python
(SaveState(saveat_ts_index=i32[], ts=f32[4], ys=f32[4,1], save_index=i32[]), SaveState(saveat_ts_index=i32[], ts=f32[4], ys=f32[4,1], save_index=i32[]))
```
The issue is that for `final_state.save_state[0]`, we want to fill in `ys` and `ts` starting at index 0, whereas for `final_state.save_state[1]` we want to start filling in at index 1 since `saveat.subs[1].t0==True`. Its this dependence on the value of `saveat_ts_index` or `save_index` that made me implement this in a loop and makes it non-obvious to me how a `tree_map` would be able to handle that.

If you agree, then I will try to refactor this as a loop-of-a-tree-map vs. the way I have it now. Very possible I am missing something simple though :)

Reply:
Hmm. I think the only cases that affect this are `t0` and `t1`? If we're saving neither of them then we should fill in `ts == t0`-many elements of the output. If we're saving just one of them then we want to save one more value. If we're saving both of them then we want to save two more values.

So untested but something like this:
```python
mask = ts == t0
if t0 or t1:
    if t0 and t1:
        mask = lax.dynamic_update_slice_in_dim(mask, jnp.array([True, True]), jnp.argmin(mask), axis=0)
    else:
        mask = lax.dynamic_update_slice_in_dim(mask, jnp.array([True]), jnp.argmin(mask), axis=0)
ys = jnp.where(mask, y0, jnp.inf)
```
?